### PR TITLE
Add MAGI-1: Autoregressive Video Generation at Scale

### LINKED ISSUE #11519: Request support for MAGI-1
### Model/Pipeline/Scheduler description

MAGI-1 is a video generation model that has achieved stunning visual effects.

### Open source status

- [x] The model implementation is available.
- [x] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

https://github.com/SandAI-org/MAGI-1
https://github.com/SandAI-org/MagiAttention

### PR DESCRIPTION
**Pull request by lavinal712**
To fix https://github.com/huggingface/diffusers/issues/11519

Paper: https://arxiv.org/abs/2505.13211
Code: https://github.com/SandAI-org/MAGI-1
Model: https://huggingface.co/sand-ai/MAGI-1

Thanks to @tolgacangoz and everyone who contributed to or reviewed the earlier work.

This PR adds initial MAGI-1 support to Diffusers, with the non-quantized MAGI-1 4.5B base checkpoint as the primary validated configuration.

It includes:

- `AutoencoderKLMagi` for causal video encoding and tiled decoding.
- `MagiTransformer3DModel` with MAGI attention, rotary embeddings, timestep conditioning, and KV-cache support.
- `MagiTextConditioningModel` for the learned null-caption, quality, and duration features.
- `MagiEulerScheduler` and the MAGI three-way classifier-free guider.
- Modular text-to-video, image-to-video, and video-to-video workflows.
- Chunked autoregressive generation and clean-prefix KV-cache management.
- A conversion script for converting the official MAGI-1 checkpoint into Diffusers format.
- Documentation, inference examples, and unit tests.

### Validation

The implementation was checked using the converted official 4.5B base weights.

This PR is still under active development and is expected to require a few more weeks of work before it is ready for final review. The implementation and validation are progressing steadily, and the remaining issues are being addressed.

Thank you for your patience and for any feedback during this process.

@yiyixuxu @dg845 and @asomoza

### CODE REVIEW

**Comment by lavinal712** [comment]
### Pipeline usage

```python
import torch

from diffusers import ComponentsManager, MagiModularPipeline
from diffusers.utils import export_to_video

manager = ComponentsManager()
manager.enable_auto_cpu_offload(device="cuda:0")

pipe = MagiModularPipeline.from_pretrained(
    "/path/to/MAGI-1-diffusers",
    components_manager=manager,
)
pipe.load_components(
    dtype={
        "default": torch.float32,
        "transformer": torch.bfloat16,
        "vae": torch.bfloat16,
    }
)

pipe.transformer.set_attention_backend("flash_varlen")
pipe.vae.set_attention_backend("flash")
pipe.vae.enable_tiling(tile_sample_min_length=12)

video = pipe(
    prompt="A golden retriever runs through a sunlit meadow.",
    height=512,
    width=512,
    num_frames=96,
    num_inference_steps=64,
    cache_device="cpu",
    generator=torch.Generator("cuda:0").manual_seed(42),
    output_type="np",
    output="videos",
)

export_to_video(video[0], "magi.mp4", fps=24)
```

**Comment by lavinal712** [comment]
https://github.com/user-attachments/assets/32d4a57a-c913-41ad-bc99-337a03fa569c
